tools/autograd/gen_python_functions.py

```diff
@@ -293,436 +158,546 @@ def get_py_torch_functions(declarations):
+def unpack_optional_dimname_list_hack(name, expr):
+    # optional<ArrayRef<T>> are special. The PythonArgParser returns an
+    # optional<vector<T>>, which cannot be implicitly converted to
+    # optional<ArrayRef<T>>. One needs to unwrap the optional and rewrap.
+    result = """\
+        auto __{name} = {expr};
+        c10::optional<{typ}> {name} = __{name} ? c10::make_optional({typ}(__{name}.value())) : c10::nullopt;
+    """.format(name=name, expr=expr, typ='DimnameList')
+    return [line.strip() for line in result.split('\n')]
+
+
+def parse_arg(arg, arg_index, unpack_to_local=False):
+    # get parsed rhs
+    expr = parsed_arg_expr(arg, arg_index)
+
+    # maybe unpack to local
+    name = arg['name']
+    typename = arg['type']
+    if typename == 'c10::optional<DimnameList>':
+        inits = unpack_optional_dimname_list_hack(name, expr)
+        expr = name
+    elif unpack_to_local:
+        inits = ['auto {} = {};'.format(name, expr)]
+        expr = name
+    else:
+        inits = []
+
+    return expr, inits
+
+
+#
+# schema type to cpp type conversions
+# some of these are to prevent dangling refs to temps, others are more obscure
+# TODO don't know if these fold into more general conversions somehere, hope so
+#
+
+TEMP_SAFE_CPP_DECL_TYPE = {
+    'Tensor': 'const Tensor &',
+    'Tensor &': 'Tensor',
+}
+
+def get_cpp_decl_type(typename, ensure_temp_safe=True):
+    if ensure_temp_safe:
+        typename = TEMP_SAFE_CPP_DECL_TYPE.get(typename, typename)
+    return typename
+
+
+def get_cpp_formal(arg, ensure_temp_safe=True):
+    decl_type = get_cpp_decl_type(arg['type'], ensure_temp_safe)
+    return '{} {}'.format(decl_type, arg['name'])
+
+
+# XXX: if you got here because of an assertion failure, it doesn't mean
+# it's enough to just extend the list here. Before you do this, make sure
+# to add an appropriate wrap() overload in torch/csrc/autograd/utils/wrap_outputs.h.
+SUPPORTED_RETURN_TYPES = {
+    'Tensor',
+    'std::tuple<Tensor,Tensor>',
+    'std::tuple<Tensor,Tensor,Tensor>',
+    'std::tuple<Tensor,Tensor,Tensor,Tensor>',
+    'std::tuple<Tensor,Tensor,Tensor,Tensor,Tensor>',
+    'std::tuple<Tensor,Tensor,Tensor,int64_t>',
+    'std::tuple<Tensor,Tensor,double,int64_t>',
+    'std::tuple<Tensor,Tensor,Tensor,Tensor,int64_t>',
+    'std::tuple<Tensor,Tensor,double,Tensor,int64_t>',
+    'std::vector<Tensor>',
+    'Scalar', 'bool', 'int64_t', 'void*', 'void',
+    'QScheme', 'double',
+    'IntArrayRef',
+    'ScalarType'
+}
+
+def get_simple_return_type(declaration):
+    # Use the simple_return_type (Tensor) rather than the fancy return type
+    # (Tensor &).  This is important because the dispatch functions take
+    # mutable arguments *by value*, not by reference.  If you then return
+    # a a reference to such an argument, you will now have a pointer to a
+    # dangling stack entry.  Not good.
+    #
+    # You want:
+    #
+    #   Tensor dispatch_selu_(Tensor self) { return at::selu_(self); }
```

Comment:
are you sure your dispatch lambda follows this pattern?  I just ask because the generated code doesn't use the explicit return syntax, so are you sure it's never inferred to be "Tensor&"?

Reply:
at a minimum, you should probably update the comment to explain the situation after your changes.

Reply:
Thanks a bunch for catching this.
- added explicit return type for the lambdas - absolutely should've done this before, zero reason to omit it. Just an oversight. 
- updated the comment, again totally should have done this.